crates/uv-resolver/src/candidate_selector.rs

```diff
@@ -89,35 +82,52 @@ impl CandidateSelector {
         preferences: &'a Preferences,
         installed_packages: &'a InstalledPackages,
         exclusions: &'a Exclusions,
+        markers: &ResolverMarkers,
     ) -> Option<Candidate<'a>> {
-        if let Some(preferred) = Self::get_preferred(
+        if let Some(preferred) = self.get_preferred(
             package_name,
             range,
             version_maps,
             preferences,
             installed_packages,
             exclusions,
+            markers,
         ) {
             return Some(preferred);
         }
 
-        self.select_no_preference(package_name, range, version_maps)
+        self.select_no_preference(package_name, range, version_maps, markers)
     }
 
     /// Get a preferred version if one exists. This is the preference from a lockfile or a locally
     /// installed version.
     fn get_preferred<'a, InstalledPackages: InstalledPackagesProvider>(
+        &self,
         package_name: &'a PackageName,
         range: &Range<Version>,
         version_maps: &'a [VersionMap],
         preferences: &'a Preferences,
         installed_packages: &'a InstalledPackages,
         exclusions: &'a Exclusions,
+        markers: &ResolverMarkers,
     ) -> Option<Candidate<'a>> {
         // If the package has a preference (e.g., an existing version from an existing lockfile),
         // and the preference satisfies the current range, use that.
         if let Some(version) = preferences.version(package_name) {
-            if range.contains(version) {
+            'preference: {
+                // Respect the version range for this requirement.
+                if !range.contains(version) {
+                    break 'preference;
+                }
+
+                // Respect the pre-release strategy for this fork.
+                if version.any_prerelease()
+                    && self.prerelease_strategy.allows(package_name, markers)
+                        != AllowPreRelease::Yes
+                {
+                    break 'preference;
+                }
```

Comment:
Is this piece correct? I feel like we should still be respecting preferences here.

Reply:
I'm not sure, I think the issue comes from [when we add preferences during resolution](https://github.com/astral-sh/uv/blob/main/crates/uv-resolver/src/resolver/mod.rs#L384), because we can end up adding a preference for a prerelease from a different fork. Maybe the solution is to make `Preferences` a `ForkMap`?

Reply:
Actually I feel like this is correct. If pre-release versions are now allowed then the version range is not actually satisfied, right?